### PR TITLE
Groups api

### LINKED ISSUE #673: Create Doing/Completed Mobile Commons Groups for a DS Campaign Run
Refs https://github.com/DoSomething/gambit/pull/672#issuecomment-255138491

When a new Campaign is cached, create [2 new Mobile Commons Groups](https://mobilecommons.zendesk.com/hc/en-us/articles/202052534-REST-API#createGroup) if they don't exist already with names
*  `campaign_[id]_run_[id]_doing` -- stores Users who haven't completed the Campaign yet
* `campaign_[id]_run_[id]_completed` -- stores Users who have completed the Campaign

We'll use these Mobile Commons Groups to segment our Mobile Commons broadcast messaging.

When a Gambit Signup document is created, add the User to the `signedup` Mobile Commons Group.
When a Gambit Signup is updated (which could happen from a POST #642) and the User has reported back, remove them from the `signedup` MC Group and add them to the `completed` Group.

### PR DESCRIPTION
**Pull request by deadlybutter**
#### What's this PR do?
- Adds gambit groups API lib
- Makes calls to Gambit Groups API. **Requires dot env file update!**
- Creates groups for campaigns if not present on the campaign

#### How should this be reviewed?
Good lord. Heres the TLDR, it works:
screen shot 2016-10-28 at 1 28 17 pm
screen shot 2016-10-28 at 1 28 24 pm

If you want to test this 100% through, you need to be really careful.
1. Delete the campaigns in your local DB.
2. Delete the group in the gambit groups prod DB.
3. Search mobile commons groups on there site for `campaign_id=some id`
4. Keep that tab open & open all 6 results into a new tab.
5. Rename each of those groups with some added random text. (We should probably start doing deleteme or some string that groups them together, sorry)
6. In that tab from step 3, select all of the groups & delete all.
7. Run Gambit.
HAVE FUN!

#### Any background context you want to provide?
I need a cookie if you're gonna make me write this one out again.

#### Relevant tickets
Fixes #673 

#### Checklist
- [ ] Tested on staging.

### CODE REVIEW

File: app/models/Campaign.js [hunk 1]
```diff
@@ -112,18 +94,47 @@ campaignSchema.statics.lookupByIDs = function (campaignIDs) {
...
-
-  // Create mobile commons group with custom name based on this campaign
-  mobilecommons.createGroup(`${prefix} status=doing`)
-  .then(doingGroup => setMobileCommonsGroup(campaign, 'doing', doingGroup))
-  .then(() => mobilecommons.createGroup(`${prefix} status=completed`))
-  .then(completedGroup => setMobileCommonsGroup(campaign, 'completed', completedGroup))
-  .then(() => campaign.save())
-  .catch(err => logger.error(err));
+  const campaignId = this._id;
+  const campaignRunId = this.current_run;
```
Comment (aaronschachter) [inline app/models/Campaign.js]: Minor question -- is it more common in JS to use `whateverId` over `whateverID`, or just a preference thing? I've been using the capital `IDs` -- i'm happy to Replace All the ID's in a future PR but let's pick one as rule to keep consistent.

**Comment by aaronschachter** [comment]
@deadlybutter --I think Gateway is looking for separate `mobilecommons_group_[status]` properties - https://github.com/DoSomething/gambit/issues/673#issuecomment-256251277

**Comment by aaronschachter** [comment]
There's also a few tasks in that comment in #673 it'd make sense to split out into more issues (like a staging script)

File: app/models/Campaign.js (hunk 1 above)
Comment (deadlybutter) [inline app/models/Campaign.js]: I think we generally do `objectId`, heres an example i randomly found in Phoenix https://github.com/DoSomething/phoenix/blob/dev/lib/themes/dosomething/paraneue_dosomething/js/components/ReactionsList.js#L100

**Comment by deadlybutter** [comment]
@aaronschachter hrm. I guess I can rename them but I would like to talk about grouping variables together to make the documents cleaner.

**Comment by deadlybutter** [comment]
@aaronschachter updated & retested

File: app/models/Campaign.js [hunk 2]
```diff
@@ -112,18 +92,47 @@ campaignSchema.statics.lookupByIDs = function (campaignIDs) {
...
+    const parsedRes = parseApiRes(res);
+    if (parsedRes) {
+      this.mobilecommons_group_doing = parsedRes.doing;
+      this.mobilecommons_group_completed = parsedRes.completed;
+      return undefined;
+    }
+
+    return gambitGroups.createGroup(campaignId, campaignRunId);
+  })
+  .then(res => {
```
Comment (sergii-tkachenko) [inline app/models/Campaign.js]: Parentheses.

File: app/models/Campaign.js [hunk 3]
```diff
@@ -112,18 +92,47 @@ campaignSchema.statics.lookupByIDs = function (campaignIDs) {
...
+    if (res === undefined) {
```
Comment (sergii-tkachenko) [inline app/models/Campaign.js]: `if (!res)`?

File: app/models/Campaign.js [hunk 4]
```diff
@@ -112,18 +92,47 @@ campaignSchema.statics.lookupByIDs = function (campaignIDs) {
...
-  .catch(err => logger.error(err));
+  const campaignId = this._id;
+  const campaignRunId = this.current_run;
+
+  function parseApiRes(res) {
+    if (res._id) {
+      // If a group is in the response, save the data for this env.
+      const env = process.env.NODE_ENV;
+      return {
+        doing: res.mobilecommons_groups[env].doing,
```
Comment (sergii-tkachenko) [inline app/models/Campaign.js]: And what if it `env` isn't in `res.mobilecommons_groups`?

File: lib/groups.js [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+const apiKey = process.env.GAMBIT_GROUPS_API_KEY;
+const apiHeaderName = 'x-gambit-group-api-key';
+
+/**
+ * Make a GET request to Gambit Groups API.
+ * Handles authentication & URL formation.
+ * @param  {string} path API Path. eg: /group
+ * @return {Promise}
+ */
+function get(path) {
```
Comment (sergii-tkachenko) [inline lib/groups.js]: Why is it outside of `module.exports`?

File: app/models/Campaign.js (hunk 4 above)
Comment (deadlybutter) [inline app/models/Campaign.js]: :thinking: that would be bad for a lot of reason! 

but I can do a safety check

File: app/models/Campaign.js (hunk 3 above)
Comment (deadlybutter) [inline app/models/Campaign.js]: sure

File: lib/groups.js (hunk 5 above)
Comment (deadlybutter) [inline lib/groups.js]: I guess we can expose it, but I didn't see a need for it

**Comment by deadlybutter** [comment]
retested again to make sure

File: lib/groups.js (hunk 5 above)
Comment (sergii-tkachenko) [inline lib/groups.js]: You may want to call it from other module in this lib. Anyway, I see no reason to explicitly make them private.

File: app/models/Campaign.js [hunk 6]
```diff
@@ -112,18 +92,48 @@ campaignSchema.statics.lookupByIDs = function (campaignIDs) {
...
-  .then(doingGroup => setMobileCommonsGroup(campaign, 'doing', doingGroup))
-  .then(() => mobilecommons.createGroup(`${prefix} status=completed`))
-  .then(completedGroup => setMobileCommonsGroup(campaign, 'completed', completedGroup))
-  .then(() => campaign.save())
-  .catch(err => logger.error(err));
+  const campaignId = this._id;
+  const campaignRunId = this.current_run;
+
+  function parseApiRes(res) {
+    if (res._id) {
```
Comment (sergii-tkachenko) [inline app/models/Campaign.js]:
```js
if (!res._id) {
  return false;
}

const env = process.env.NODE_ENV;
if (!res.mobilecommons_groups[env]) {
  return false;
}

return {
  doing: res.mobilecommons_groups[env].doing,
  completed: res.mobilecommons_groups[env].completed,
}
```

Comment (deadlybutter) [inline app/models/Campaign.js]: yeah that works 👍

File: lib/groups.js [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+ * Handles authentication & URL formation.
+ * @param  {string} path API Path. eg: /group
+ * @param  {object} data JSON data to send.
+ * @return {Promise}
+ */
+function post(path, data) {
+  return superagent.post(`${uri}${path}`)
+    .send(data)
+    .set(apiHeaderName, apiKey)
+    .then(res => res.body)
```
Comment (sergii-tkachenko) [inline lib/groups.js]: Wait. Why do you call then/catch here? Just return the promise. You're thenning it above anyways.

Comment (deadlybutter) [inline lib/groups.js]: If I do it here I only have to access res.body once and handle the error once.

Comment (deadlybutter) [inline lib/groups.js]: So, both `findGroup` and `createGroup` would need there own response + error handlers. Which doesn't seem very dry to me

Comment (sergii-tkachenko) [inline lib/groups.js]: That's actually a bad idea to handle the error here. For example, you just want to handle different errors in different manners, like retry on timeout, exit on 500.

Comment (sergii-tkachenko) [inline lib/groups.js]: And regarding handling `res.body`: what if you wanted to check response headers in user code? Or HTTP?

Comment (deadlybutter) [inline lib/groups.js]: see my reply above ^ -- complications I have no reason to worry about right now

Comment (sergii-tkachenko) [inline lib/groups.js]: ok

File: lib/groups.js [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+    return post('/group', {
+      campaign_id: campaignId,
+      campaign_run_id: campaignRunId
+    });
+  },
+
+  /**
+   * See docblock above.
+   */
+  get: get,
```
Comment (sergii-tkachenko) [inline lib/groups.js]: Why not just move func body here?

Comment (deadlybutter) [inline lib/groups.js]: was being lazy, sure i can move the entire thing

**Comment by deadlybutter** [comment]
all tests passed again, any final thoughts on this @sergii-tkachenko ?

**Pull request merged**